pkg/oauthserver/oauth/external/gitlab/gitlab.go

```diff
@@ -20,19 +20,26 @@ const (
 	// Uses the GitLab User-API (http://doc.gitlab.com/ce/api/users.html#current-user)
 	// and OAuth-Provider (http://doc.gitlab.com/ce/integration/oauth_provider.html)
 	// with default OAuth scope (http://doc.gitlab.com/ce/api/users.html#current-user)
-	// Requires GitLab 7.7.0 or higher
+	// Requires GitLab 7.7.0 or higher (released prior to 2015-09-22, we support it in OS v3.1+)
 	gitlabAuthorizePath = "/oauth/authorize"
 	gitlabTokenPath     = "/oauth/token"
-	gitlabUserAPIPath   = "/api/v3/user"
 	gitlabOAuthScope    = "api"
+
+	// the only thing different about the APIs is the endpoint
+	// the JSON data is the same
+	// v3 was removed in GitLab 11 2018-06-22 (deprecated in GitLab 9.5 2017-08-22)
+	gitlabUserAPIPathV3 = "/api/v3/user"
+	// v4 was added in GitLab 9 2017-03-22
+	gitlabUserAPIPathV4 = "/api/v4/user"
 )
 
 type provider struct {
 	providerName string
 	transport    http.RoundTripper
 	authorizeURL string
 	tokenURL     string
-	userAPIURL   string
+	userAPIURLV3 string
+	userAPIURLV4 string
```

Comment:
I think we may want to keep just userAPIURL and populate it with V4 endpoint.
On the first attempt if we get a not found error we check if the URL is the V4 and we switch it to V3 and retry.
I do not think it makes sense to keep falling back at every request.

Reply:
That will require adding some form of locking.

Reply:
or an atomic holding an index to a 2 elements array with the two strings